OneSignalSDK/onesignal/src/main/java/com/onesignal/OutcomeEventsController.java

```diff
@@ -100,32 +126,56 @@ void onSuccess(String response) {
-        if (eventsSent.contains(name)) {
-            //Event already sent
-            return;
+        final JSONArray notificationIds = osSessionManager.getSessionResult().notificationIds;
+
+        // Special handling for unique outcomes in the attributed and unattributed scenarios
+        if (osSessionManager.getSession().isAttributed()) {
+            // Make sure unique notificationIds exist before trying to make measure request
+            final JSONArray uniqueNotificationIds = getUniqueNotificationIds(name, notificationIds);
+            if (uniqueNotificationIds == null) {
+                OneSignal.Log(OneSignal.LOG_LEVEL.DEBUG,
+                        "Measure endpoint will not send because unique outcome already sent for: " +
+                                "\nSession: " + osSessionManager.getSession().toString() +
+                                "\nOutcome name: " + name +
+                                "\nnotificationIds: " + notificationIds);
+                return;
+            }
+
+            sendOutcomeEvent(name, uniqueNotificationIds, null, callback);
+
+        } else if (osSessionManager.getSession().isUnattributed()) {
+            // Make sure unique outcome has not been sent for current unattributed session
+            if (unattributedUniqueOutcomeEventsSentSet.contains(name)) {
+                OneSignal.Log(OneSignal.LOG_LEVEL.DEBUG,
+                        "Measure endpoint will not send because unique outcome already sent for: " +
+                                "\nSession: " + osSessionManager.getSession().toString() +
+                                "\nOutcome name: " + name);
+                return;
+            }
+
+            unattributedUniqueOutcomeEventsSentSet.add(name);
+            sendOutcomeEvent(name, null, null, callback);
         }
-        sendOutcomeEvent(name, callback);
-        //Even if the outcome fail we don't attempt to send it again because we have cache
-        eventsSent.add(name);
     }
 
     void sendOutcomeEventWithValue(@NonNull String name, float value, @Nullable final OneSignal.OutcomeCallback callback) {
+        final JSONArray notificationIds = osSessionManager.getSessionResult().notificationIds;
         OutcomeParams params = OutcomeParams.Builder
                 .newInstance()
                 .setWeight(value)
                 .build();
-        sendOutcomeEvent(name, params, callback);
+        sendOutcomeEvent(name, notificationIds, params, callback);
     }
 
-    private void sendOutcomeEvent(@NonNull final String name, @Nullable final OutcomeParams params, @Nullable final OneSignal.OutcomeCallback callback) {
+    private void sendOutcomeEvent(@NonNull final String name, @Nullable final JSONArray notificationIds, @Nullable final OutcomeParams params, @Nullable final OneSignal.OutcomeCallback callback) {
         OSSessionManager.SessionResult sessionResult = osSessionManager.getSessionResult();
```

Comment:
no need to call this again right? you can pass a new OSSessionManager.SessionResult

Reply:
I have to modify `notificationIds` in the ATTRIBUTED for unique outcome events

Reply:
yeah i mean intead of passing JSONArray notificationIds as param,  pass an OSSessionManager.SessionResult in that way you are not calling again osSessionManager.getSessionResult()

Reply:
We cant modify the `notificationIds` inside of the `SessionResult` so we would have to pass another param then

Reply:
Should we pass the extra param? It will be a 5 param method then

Reply:
but you can build a new SessionResult. In that way getSessionResult is not being called twice

Reply:
Running `getSessionResult()` more than once won't have any side effects. I think we might be able to clean this up with some kind of internal `OutcomeEvent` class in the future instead of adding more params. But I don't think it is necessary to make this change now.